### PR TITLE
revset: prep for migrating parser and alias substition function

### PR DESCRIPTION
**Pull request by yuja**
# Checklist

If applicable:
- [ ] I have updated `CHANGELOG.md`
- [ ] I have updated the documentation (README.md, docs/, demos/)
- [ ] I have updated the config schema (cli/src/config-schema.json)
- [x] I have added tests to cover my changes

### CODE REVIEW

File: lib/src/dsl_util.rs [hunk 1]
```diff
@@ -105,11 +119,94 @@ impl<'i, T> FunctionCallNode<'i, T> {
...
+        InvalidArguments<'i>,
+    > {
+        if self.keyword_args.is_empty() {
+            let (required, optional) = self.expect_arguments::<N, M>()?;
+            // TODO: use .each_ref() if MSRV is bumped to 1.77.0
+            Ok((array::from_fn(|i| &required[i]), optional))
+        } else {
+            let (required, optional) = self.expect_named_arguments_vec(names, N, N + M)?;
+            Ok((
+                required.try_into().map_err(|_: Vec<_>| ()).unwrap(),
```
Comment (torquestomp) [inline lib/src/dsl_util.rs:145]: Can this be `.ok()` instead of `.map_err(consume_error)`?

File: lib/src/dsl_util.rs [hunk 2]
```diff
@@ -503,3 +616,92 @@ where
...
+        assert_eq!(
+            f.expect_some_arguments::<0>().unwrap(),
+            (&[], [value(0)].as_slice())
+        );
+        assert_eq!(
+            f.expect_some_arguments::<1>().unwrap(),
+            (&[value(0)], [].as_slice())
+        );
+        assert!(f.expect_arguments::<0, 0>().is_err());
+        assert_eq!(f.expect_arguments::<1, 1>().unwrap(), (&[value(0)], [None]));
```
Comment (torquestomp) [inline lib/src/dsl_util.rs:673]: Missing tests for <0, 1>

File: lib/src/dsl_util.rs [hunk 3]
```diff
@@ -503,3 +616,92 @@ where
...
+        );
+        assert_eq!(
+            f.expect_named_arguments::<1, 1>(&["a", "b"]).unwrap(),
+            ([&value(0)], [None])
+        );
+        assert!(f.expect_named_arguments::<1, 1>(&["b", "a"]).is_err());
+
+        let f = function("foo", [value(0)], [keyword("a", 1), keyword("b", 2)]);
+        assert!(f.expect_named_arguments::<0, 0>(&[]).is_err());
+        assert!(f.expect_named_arguments::<1, 1>(&["a", "b"]).is_err());
```
Comment (torquestomp) [inline lib/src/dsl_util.rs:698]: Can we add a test for ["b", "a"] (e.g. input order does not match programmer order)?  Is this expected to permute the input spans to match the specified keyword order?

**Review by torquestomp — CHANGES_REQUESTED** [review]

File: lib/src/dsl_util.rs (hunk 1 above)
Comment (yuja) [reply lib/src/dsl_util.rs:145]: `.ok()` also works. Inserted another commit to replace existing `.map_err()`.

File: lib/src/dsl_util.rs (hunk 2 above)
Comment (yuja) [reply lib/src/dsl_util.rs:673]: Added

File: lib/src/dsl_util.rs (hunk 3 above)
Comment (yuja) [reply lib/src/dsl_util.rs:698]: I think you mean `[_, "b", "a"]`. Added tests. The order of the returned array should follow the `names` arguments.

**Review by torquestomp — APPROVED** [review]

**Pull request merged**